### PR TITLE
Updates to OpenThread support

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description

Latest updates to Thread support

- Remove support for defining OpenThread configuration at build time. We always need the managed code to configure the support.  We may later support the saving of config so that it can automatically started on boot.
-  Add extra support in build preset for C6 & H2
-  ESP32_PSRAM_REV3_IPV6
-- Update Readme
-- Add to nightly build
-- Add Thread support 
- Allow a network interface to start later (after boot)
-- Add SetSocketDriverInterface method in LWIP_SOCKETS to allow driver interface number to be set after start.
- Add GetNativeIPV6Option() to lwip support
- Add new event EVENT_OPENTHREAD  (ALready done in Events lib)
- Update partitions table for C6
- Add support for CLI output from OpenThread  and support startup later from code.
- Tidy up code

## Motivation and Context

Latest changes & fixes for IDF 5 release

## How Has This Been Tested?

Sample code to be released.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies/declarations (update dependencies or assembly declarations and changes associated, has no impact on code or features)
- [x] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

## Summary by CodeRabbit

- **New Features**
  - Introduced a unified `THREAD_DEVICE_TYPE` option to simplify OpenThread configuration.
  - Enabled additional APIs for system devices and I/O operations on ESP32 platforms.
  - Added support for new native methods and functionalities related to OpenThread.
  - Included a new target `ESP32_PSRAM_REV3_IPV6` for enhanced IPv6 and Thread support.
  
- **Bug Fixes**
  - Improved handling of IPv6 options in socket operations.

- **Documentation**
  - Updated `README.md` to include the new target `ESP32_PSRAM_REV3_IPV6`.

- **Refactor**
  - Reorganized OpenThread configuration options to streamline device type specification.

- **Chores**
  - Adjusted Azure Pipelines configuration to include the new ESP32 target.

### CODE REVIEW

File: targets/ESP32/_nanoCLR/nanoFramework.Networking.Thread/net_thread_native_nanoFramework_Networking_Thread_OpenThread.cpp [hunk 1]
```diff
@@ -93,26 +79,102 @@
...
+                OT_NETWORK_KEY_SIZE);
+
+            inUse |= OpenThreadDataset_componentInUseFlags_networkKey;
+        }
+
+        // Save in use flags
+        datasetObject[OpenThreadDataSet::FIELD___inUseFlags].NumericByRef().u4 = inUse;
+    }
+    NANOCLR_NOCLEANUP();
+}
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/nanoFramework.Networking.Thread/net_thread_native_nanoFramework_Networking_Thread_OpenThread.cpp:154]: no resources to cleanup

File: src/PAL/Lwip/lwIP_Sockets.cpp [hunk 2]
```diff
@@ -1159,23 +1164,23 @@
...
 
     if (from)
     {
         Sock_SockaddrToSockaddr(from, (sockaddr *)&addr, fromlen);
-
-        pFrom = (sockaddr *)&addr;
     }
 
-    ret = lwip_recvfrom(socket, buf, len, flags, pFrom, (u32_t *)fromlen);
+    ret = lwip_recvfrom(socket, buf, len, flags, (sockaddr *)&addr, &addrLen);
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply src/PAL/Lwip/lwIP_Sockets.cpp:1183]: Already handled

File: targets/ESP32/_nanoCLR/nanoFramework.Networking.Thread/net_thread_native.h [hunk 3]
```diff
@@ -44,6 +46,17 @@ typedef enum __nfpack ThreadDeviceRole
...
+typedef enum __nfpack ThreadDeviceState
+{
+    ThreadDeviceState_OPENTHREAD_EVENT_START = 0,
+    ThreadDeviceState_OPENTHREAD_EVENT_STOP = 1,
+    ThreadDeviceState_OPENTHREAD_EVENT_DETACHED = 2,
+    ThreadDeviceState_OPENTHREAD_EVENT_ATTACHED = 3,
+    ThreadDeviceState_OPENTHREAD_EVENT_IF_UP = 5,
+    ThreadDeviceState_OPENTHREAD_EVENT_IF_DOWN = 6,
+    ThreadDeviceState_OPENTHREAD_EVENT_GOT_IP6 = 7,
+} ThreadDeviceState;
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/nanoFramework.Networking.Thread/net_thread_native.h:58]: Coming from C#

File: targets/ESP32/_nanoCLR/nanoFramework.Networking.Thread/net_thread_native_nanoFramework_Networking_Thread_OpenThread.cpp [hunk 4]
```diff
@@ -127,54 +189,184 @@
...
+            oterr = otDatasetGetActive(instance, &dataset);
+            if (oterr != OT_ERROR_NONE)
+            {
+                NANOCLR_SET_AND_LEAVE(CLR_E_ENTRY_NOT_FOUND);
+            }
+            StoreDataset(stack, dataset, datasetRef);
+        }
+    }
+    NANOCLR_NOCLEANUP();
+}
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/nanoFramework.Networking.Thread/net_thread_native_nanoFramework_Networking_Thread_OpenThread.cpp:275]: Does use it if null

File: targets/ESP32/_nanoCLR/nanoFramework.Networking.Thread/net_thread_native_nanoFramework_Networking_Thread_OpenThread.cpp [hunk 5]
```diff
@@ -127,54 +189,184 @@
...
-        oterr = otDatasetSetActive(instance, &dataset);
-        if (oterr !=  OT_ERROR_NONE)
+        // Bring up IPV6 interface
+        oterr = otIp6SetEnabled(instance, true);
+        if (oterr != OT_ERROR_NONE)
         {
-        debug_printf("otDatasetSetActive fails %d", oterr);
             NANOCLR_SET_AND_LEAVE(CLR_E_FAIL);
-        } 
+        }
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/nanoFramework.Networking.Thread/net_thread_native_nanoFramework_Networking_Thread_OpenThread.cpp:207]: Uses NANOCLR_SET_AND_LEAVE to exit function

**Comment by AdrianSoundy** [comment]
@josesimoes 
Added you for review as it's a lot of changes but mainly just for OpenThread which you can probably ignore for now.

The only change in Network code is to allow a Network interface to fail at boot and to continue with next interface
Also adding a way to start network interface later once managed code running by setting network interface number in global structure.

**Review by josesimoes — APPROVED** [review]
LGTM!
I presume you've tested all this, in particular, the changes related with the network interface. Let's go for it.

**Pull request merged**